prelude/runtime/string.c

```diff
@@ -37,15 +37,20 @@ SkipInt SKIP_String_hash(char* obj) {
   return (SkipInt)str->hash;
 }
 
-// Allocates a string (with no data nor hash).
+// Allocates and nul-terminates a string (with no data nor hash).
 char* sk_string_alloc(uint32_t size) {
   sk_string_t* obj =
-      (sk_string_t*)SKIP_Obstack_alloc(size + sk_string_header_size);
+      (sk_string_t*)SKIP_Obstack_alloc(sk_string_header_size + size + 1);
   obj->size = size;
-  return (char*)&(obj->data);
+  char* str = (char*)&(obj->data);
+  str[size] = '\0';
+  return str;
 }
 
 char* sk_string_create(const char* buffer, uint32_t size) {
+  if (size > 0 && buffer[size - 1] == '\0') {
+    size--;
```

Comment:
This looks ugly...

And it is actually ugly because it will change what's stored in the `size` field of the resulting `sk_string_t`. Is it intended?

Reply:
The intent here is to avoid adding a nul terminator if there already is one, which will lead to the resulting string being considered to have an embedded nul character, and blow up if passed to C. It isn't pretty. I audited all the call sites, and it looks like the only way for that to happen is if creating a string from an array of uint8 where the last element is 0 or to create a string as the contents of a file where the last byte of the file is 0. Keeping the extra nul in those cases seems legit as they are part of the explicit input data. So I've updated to remove this check.